### PR TITLE
ASoC: rt722-sdca: Add RT722 SDCA driver

### PR DESCRIPTION
**Pull request by jack-cy-yu**
This is the initial codec driver for rt722-sdca.

### CODE REVIEW

File: sound/soc/codecs/Kconfig [hunk 1]
```diff
@@ -200,6 +200,7 @@ config SND_SOC_ALL_CODECS
 	imply SND_SOC_RT711_SDW
 	imply SND_SOC_RT711_SDCA_SDW
 	imply SND_SOC_RT712_SDCA_SDW
+	imply SND_SOC_RT722_SDCA_SDW
```
Comment (plbossart) [inline sound/soc/codecs/Kconfig:203]: @jack-cy-yu please use branches in your own github fork, please do NOT push branches on the SOF gitHub, thank you.

File: sound/soc/codecs/rt722-sdca.h [hunk 2]
```diff
@@ -0,0 +1,240 @@
...
+	int jack_type;
+	int jd_src;
+	unsigned int scp_sdca_stat1;
+	unsigned int scp_sdca_stat2;
+	unsigned int hw_id;
+	bool fu1e_dapm_mute;
+	bool fu0f_dapm_mute;
+	bool fu0f_mixer_l_mute;
+	bool fu0f_mixer_r_mute;
+	bool fu1e_mixer_mute[4];
```
Comment (plbossart) [inline sound/soc/codecs/rt722-sdca.h:42]: maybe group the members by function, e.g. all jack stuff together, all amp stuff together and all 'generic' stuff also separated.

File: sound/soc/codecs/rt722-sdca.h [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+
+struct rt722_sdca_dmic_kctrl_priv {
+	unsigned int reg_base;
+	unsigned int count;
+	unsigned int max;
+	unsigned int invert;
+};
+
+struct sdw_stream_data {
+	struct sdw_stream_runtime *sdw_stream;
```
Comment (plbossart) [inline sound/soc/codecs/rt722-sdca.h:53]: no, we removed this recently, please see https://lore.kernel.org/alsa-devel/168018290301.3345013.10279767218349232825.b4-ty@kernel.org/

File: sound/soc/codecs/rt722-sdca.h [hunk 4]
```diff
@@ -0,0 +1,240 @@
...
+#define RT722_EAPD_CTL						0x55
+#define RT722_UMP_HID_CTL0					0x60
+#define RT722_UMP_HID_CTL1					0x61
+#define RT722_UMP_HID_CTL2					0x62
+#define RT722_UMP_HID_CTL3					0x63
+#define RT722_UMP_HID_CTL4					0x64
+#define RT722_UMP_HID_CTL5					0x65
+#define RT722_UMP_HID_CTL6					0x66
+#define RT722_UMP_HID_CTL7					0x67
+#define RT722_UMP_HID_CTL8					0x68
```
Comment (plbossart) [inline sound/soc/codecs/rt722-sdca.h:133]: the tab alignment is difficult to follow, it's not clear if the different indent is intentional or not.

File: sound/soc/codecs/rt722-sdca.h [hunk 5]
```diff
@@ -0,0 +1,240 @@
...
+#define RT722_SDCA_RATE_32000HZ		0x07
+#define RT722_SDCA_RATE_44100HZ		0x08
+#define RT722_SDCA_RATE_48000HZ		0x09
+#define RT722_SDCA_RATE_96000HZ		0x0b
+#define RT722_SDCA_RATE_192000HZ	0x0d
+
+enum {
+	RT722_AIF1,
+	RT722_AIF2,
+	RT722_AIFS,
```
Comment (plbossart) [inline sound/soc/codecs/rt722-sdca.h:223]: which one is used by what function, it'd be good to clarify?

File: sound/soc/codecs/rt722-sdca.c [hunk 6]
```diff
@@ -0,0 +1,1632 @@
...
+
+		for (idx = 0; idx < sizeof(buf); idx++) {
+			ret = regmap_read(rt722->regmap,
+				RT722_BUF_ADDR_HID1 + offset + idx, &val);
+			if (ret < 0)
+				goto io_error;
+			buf[idx] = val & 0xff;
+		}
+
+		if (buf[0] == 0x11) {
```
Comment (plbossart) [inline sound/soc/codecs/rt722-sdca.c:292]: this looks similar to the code in https://github.com/thesofproject/linux/pull/4294/commits/6728bbcac2cef73126ca92c32259082d993171f4#diff-0784eed83ac441dcb3d4d7bdc4048ce9ecf08dcd7cbd5140e5c66825c2e0a4bfR106, is there an opportunity to use a helper and remove copy-pasted code (if this is indeed the case?)

File: sound/soc/codecs/rt722-sdca.c [hunk 7]
```diff
@@ -0,0 +1,1632 @@
...
+		if (mc->shift == 8) /* boost gain */
+			ctl_r = read_r / tendB;
+		else { /* ADC/DAC gain */
+			if (adc_vol_flag)
+				ctl_r = mc->max - (((0x1e00 - read_r) & 0xffff) / interval_offset);
+			else
+				ctl_r = mc->max - (((0 - read_r) & 0xffff) / interval_offset);
+		}
+	} else
+		ctl_r = ctl_l;
```
Comment (plbossart) [inline sound/soc/codecs/rt722-sdca.c:530]: usually we have brackets here as well, please see checkpatch results.

File: sound/soc/codecs/rt722-sdca.c [hunk 8]
```diff
@@ -0,0 +1,1632 @@
...
+	return 0;
+}
+
+static int rt722_sdca_set_fu1e_capture_ctl(struct rt722_sdca_priv *rt722)
+{
+	int err, i;
+	unsigned int ch_mute;
+
+	for (i = 0; i < ARRAY_SIZE(rt722->fu1e_mixer_mute); i++) {
+		ch_mute = (rt722->fu1e_dapm_mute || rt722->fu1e_mixer_mute[i]) ? 0x01 : 0x00;
```
Comment (plbossart) [inline sound/soc/codecs/rt722-sdca.c:544]: wouldn't this work?

ch_mute = rt722->fu1e_dapm_mute || rt722->fu1e_mixer_mute[i];

File: sound/soc/codecs/rt722-sdca.c [hunk 9]
```diff
@@ -0,0 +1,1632 @@
...
+		regmap_read(rt722->mbq_regmap, p->reg_base + i, &regvalue[i]);
+
+		gain_val[i] = ucontrol->value.integer.value[i];
+		if (gain_val[i] > p->max)
+			gain_val[i] = p->max;
+
+		if (!adc_vol_flag) /* boost gain */
+			gain_val[i] = gain_val[i] * 0x0a00;
+		else { /* ADC gain */
+			gain_val[i] = 0x1e00 - ((p->max - gain_val[i]) * interval_offset);
```
Comment (plbossart) [inline sound/soc/codecs/rt722-sdca.c:722]: lots of magic numbers in this function, can we use defines here?

File: sound/soc/codecs/rt722-sdca.c [hunk 10]
```diff
@@ -0,0 +1,1632 @@
...
+			changed = 1;
+	}
+
+	if (!changed)
+		return 0;
+
+	for (i = 0; i < p->count; i++) {
+		err = regmap_write(rt722->mbq_regmap, p->reg_base + i, gain_val[i]);
+		if (err < 0)
+			dev_err(&rt722->slave->dev, "0x%08x can't be set\n", p->reg_base + i);
```
Comment (plbossart) [inline sound/soc/codecs/rt722-sdca.c:736]: %#08x

File: sound/soc/codecs/rt722-sdca.c [hunk 11]
```diff
@@ -0,0 +1,1632 @@
...
+	.dapm_routes = rt722_sdca_audio_map,
+	.num_dapm_routes = ARRAY_SIZE(rt722_sdca_audio_map),
+	.set_jack = rt722_sdca_set_jack_detect,
+	.endianness = 1,
+};
+
+static int rt722_sdca_set_sdw_stream(struct snd_soc_dai *dai, void *sdw_stream,
+				int direction)
+{
+	struct sdw_stream_data *stream;
```
Comment (plbossart) [inline sound/soc/codecs/rt722-sdca.c:1216]: remove as done in upstream code

File: sound/soc/codecs/rt722-sdca.c [hunk 12]
```diff
@@ -0,0 +1,1632 @@
...
+	else
+		dai->capture_dma_data = stream;
+
+	return 0;
+}
+
+static void rt722_sdca_shutdown(struct snd_pcm_substream *substream,
+				struct snd_soc_dai *dai)
+{
+	struct sdw_stream_data *stream;
```
Comment (plbossart) [inline sound/soc/codecs/rt722-sdca.c:1239]: remove as done in upstream code

File: sound/soc/codecs/rt722-sdca.c [hunk 13]
```diff
@@ -0,0 +1,1632 @@
...
+
+	if (!rt722->slave)
+		return -EINVAL;
+
+	/* SoundWire specific configuration */
+	if (substream->stream == SNDRV_PCM_STREAM_PLAYBACK) {
+		direction = SDW_DATA_DIR_RX;
+		if (dai->id == RT722_AIF1)
+			port = 1;
+		else if (dai->id == RT722_AIF2)
```
Comment (plbossart) [inline sound/soc/codecs/rt722-sdca.c:1273]: we'd really need more details on which interface is which, e.g. RT722_AIF2_AMP or something

File: sound/soc/codecs/rt722-sdca.c [hunk 14]
```diff
@@ -0,0 +1,1632 @@
...
+	return 0;
+}
+
+static int rt722_sdca_pcm_hw_free(struct snd_pcm_substream *substream,
+				struct snd_soc_dai *dai)
+{
+	struct snd_soc_component *component = dai->component;
+	struct rt722_sdca_priv *rt722 = snd_soc_component_get_drvdata(component);
+	struct sdw_stream_data *stream =
+		snd_soc_dai_get_dma_data(dai, substream);
```
Comment (plbossart) [inline sound/soc/codecs/rt722-sdca.c:1356]: remove as in upstream code

File: sound/soc/codecs/rt722-sdca.c [hunk 15]
```diff
@@ -0,0 +1,1632 @@
...
+			.rates = RT722_STEREO_RATES,
+			.formats = RT722_FORMATS,
+		},
+		.capture = {
+			.stream_name = "DP6 Capture",
+			.channels_min = 1,
+			.channels_max = 2,
+			.rates = RT722_STEREO_RATES,
+			.formats = RT722_FORMATS,
+		},
```
Comment (plbossart) [inline sound/soc/codecs/rt722-sdca.c:1413]: again it's not clear which interface/port is used for what.

**Review by plbossart — CHANGES_REQUESTED** [review]
@jack-cy-yu see below suggestions to improve the code, thanks!

File: sound/soc/codecs/Kconfig (hunk 1 above)
Comment (jack-cy-yu) [reply sound/soc/codecs/Kconfig:203]: Got it. I'll re-send another PR for this driver.

File: sound/soc/codecs/rt722-sdca.h (hunk 2 above)
Comment (jack-cy-yu) [reply sound/soc/codecs/rt722-sdca.h:42]: The format like below ?

struct  rt722_sdca_priv {
	struct regmap *regmap;
	struct regmap *mbq_regmap;
	struct snd_soc_component *component;
	struct sdw_slave *slave;
	enum sdw_slave_status status;
	struct sdw_bus_params params;
	bool hw_init;
	bool first_hw_init;
	struct mutex calibrate_mutex;
	struct mutex disable_irq_lock;
	bool disable_irq;
	/* For Headset jack & Headphone */
	unsigned int scp_sdca_stat1;
	unsigned int scp_sdca_stat2;
	struct snd_soc_jack *hs_jack;
	struct delayed_work jack_detect_work;
	struct delayed_work jack_btn_check_work;
	int jack_type;
	int jd_src;
	bool fu0f_dapm_mute;
	bool fu0f_mixer_l_mute;
	bool fu0f_mixer_r_mute;
	/* For DMIC */
	bool fu1e_dapm_mute;
	bool fu1e_mixer_mute[4];
};

File: sound/soc/codecs/rt722-sdca.h (hunk 3 above)
Comment (jack-cy-yu) [reply sound/soc/codecs/rt722-sdca.h:53]: will fix it

File: sound/soc/codecs/rt722-sdca.h (hunk 4 above)
Comment (jack-cy-yu) [reply sound/soc/codecs/rt722-sdca.h:133]: will fix it

File: sound/soc/codecs/rt722-sdca.h (hunk 5 above)
Comment (jack-cy-yu) [reply sound/soc/codecs/rt722-sdca.h:223]: will fix it

File: sound/soc/codecs/rt722-sdca.c (hunk 11 above)
Comment (jack-cy-yu) [reply sound/soc/codecs/rt722-sdca.c:1216]: will fix it

File: sound/soc/codecs/rt722-sdca.c (hunk 6 above)
Comment (jack-cy-yu) [reply sound/soc/codecs/rt722-sdca.c:292]: will fix it

File: sound/soc/codecs/rt722-sdca.c (hunk 7 above)
Comment (jack-cy-yu) [reply sound/soc/codecs/rt722-sdca.c:530]: will fix it

File: sound/soc/codecs/rt722-sdca.c (hunk 8 above)
Comment (jack-cy-yu) [reply sound/soc/codecs/rt722-sdca.c:544]: will fix it

File: sound/soc/codecs/rt722-sdca.c (hunk 9 above)
Comment (jack-cy-yu) [reply sound/soc/codecs/rt722-sdca.c:722]: will fix it

File: sound/soc/codecs/rt722-sdca.c (hunk 10 above)
Comment (jack-cy-yu) [reply sound/soc/codecs/rt722-sdca.c:736]: will fix it

File: sound/soc/codecs/rt722-sdca.c (hunk 12 above)
Comment (jack-cy-yu) [reply sound/soc/codecs/rt722-sdca.c:1239]: will fix it

File: sound/soc/codecs/rt722-sdca.c (hunk 13 above)
Comment (jack-cy-yu) [reply sound/soc/codecs/rt722-sdca.c:1273]: will fix it

File: sound/soc/codecs/rt722-sdca.c (hunk 14 above)
Comment (jack-cy-yu) [reply sound/soc/codecs/rt722-sdca.c:1356]: will fix it

File: sound/soc/codecs/rt722-sdca.c (hunk 15 above)
Comment (jack-cy-yu) [reply sound/soc/codecs/rt722-sdca.c:1413]: will fix it

**Comment by jack-cy-yu** [comment]
Close this PR due to another PR (#4307) has been sent regarding to the same rt722-sdca driver.

**Pull request closed without merging**